### DISCUSSION TITLE
possibilities to adjust maximum power limit (DPL) according to current EPEX spot hourly price

### DISCUSSION THREAD

**Discussion by 1-am-r00t** [hoylabs/OpenDTU-OnBattery#1021] [Ideas]
Hi!

Ever since I switched to an EPEX Spot Hourly energy tariff in December 2023, I've been experimenting with adjusting the maximum power limit for DPL according to the current energy price. Think of it like this: highest price of the day = 100% max power limit, lowest price = 0% (actually the absolute minimum is 10% because 0% would lead to the inverter getting stuck in standby and not recoving from standby unless restarted). All the hours in between are calculated as percentage relative to min price & max price depending on the corresponding price of energy for the current hour. Like this I am able to further improve energy cost and efficiency, because say if the price for the corresponding hour drops below 0 ct, I'm not interested in "wasting" any amount of precious energy into the grid.

What I ended up doing is I've implemented in Node-Red some flows to ...:
1. fetch market data for next 24h from my energy provider's API shortly before midnight
2. calculate DPL (absolute) max power limit percentage for each hour and store it as a global context data dictionary
3. at every full hour start, read the calculated percentage from global context data, check the variable that defines the maximum power based on physics (cable diameter + cable length) - in my case I'm operating a HM-800 and my physics currently allow for no more than 510W given the fact I am using a DIY 24V battery system.
4. since no native MQTT topic exists for setting these DPL parameters, I need to perform an HTTP request to "mock" changing this setting in the GUI (= HTTP POST to http://<IP_OF_DTU>/api/powerlimiter/config)

However, changing the DPL settings via HTTP request every hour seems to be a bit cumbersome since native MQTT topics exist, just not for the maximum power limit.

So, long story short - I have two ideas, one out of which I guess could be quite easy to implement:
- MQTT cmd endpoint to set more DPL parameters - "maximum power limit" in this case (beside "only" the Solar Passthrough parameters)
- implement a possibility to have the DTU fetch the market data by itself from an EPEX Spot API, set the limit according to a logic like I mentioned above and skip the Node-Red part (nice-to-have, I'm not even sure this will be possible given the recent memory limitations experienced on non S3 ESP's) - I think this discussion about a scheduler might come in handy for my idea https://github.com/helgeerbe/OpenDTU-OnBattery/discussions/586

Happy to hear your inputs! I can imagine I am not the first guy who implemented something like this :)

**Comment by spcqike** [comment]
Sry, ich schreib einfach mal deutsch, da es einfacher ist und doch recht komplex :)

zuerst: so ganz verstehe ich deine Idee, bzw. die Intension, nicht. 

Du willst also die maximale Leistung des DPL dynamisch, in Abhängigkeit vom EPEX Preis, einstellen. 100% Leistung, wenn der Strompreis "maximal" ist, 0% (bzw, 10%) wenn er "minimal" ist. 
Weil: du willst keinen Strom "verschwenden", wenn der EPEX Preis negativ ist.

Wo genau bleibt da deine eigene Versorgung? Willst du den DPL drosseln, weil der Strom auf der Börse tagsüber günstig ist, und statt deine Grundlast selbst zudecken lieber einkaufen?

Wäre es nicht sinniger, den Zielwert des DPL anzupassen, abhängig vom Börsenpreis? Sodass die eigene Grundlast doch gedeckt wird.


Ich hatte solche Gedanken in der Vergangenheit ebenfalls. aber nicht, weil ich keinen Strom "verschwenden" oder "verschenken" will, sondern weil ich niemanden "zur Last" fallen möchte.
Aber, der Strommarkt ist viel zu komplex und für uns, Stand heute, undurchsichtig, als das wir das alles überblicken und abbilden könnten. Dafür fehlen uns einfach viel zu viele Infos.

ein negativer Börsenpreis bewirkt, dass der Anbieter dem Kunden Geld gibt, wenn er Strom abnimmt. Damit er Kraftwerke nicht drosseln oder gar stoppen muss, nimmt er halt 1-2 h "Verlust" in Kauf.

Aber: nur weil der EPEX Preis heute zu einer Stunde 0 oder gar negativ ist, heißt das nicht, dass das auch für unser jeweiliges lokales Netzsegment zutrifft.

Beispiel:
Der Wind bläst uns die Pfannen vom Dach. An der Küste ist Strom im Überfluss, negative Preise auf der Börse.
Die Kunden können also "netzdienlich" überschüssigen Strom abnehmen und werden mit niedrigen bis negativen Preisen belohnt. Das gilt dann an dem Tag so erst einmal für alle deutschen Stromkunden

Nun wohne ich aber im Süden und der Windstrom kommt gar nicht bis zu mir. Es erfolgt ein Redispatch, Windkraft wird abgeriegelt (und bekommt Entschädigung), Gaskraft wird hochgefahren (und kostet entsprechend Geld), der Verbraucher zahlt aber dennoch wenig bis nichts, ja bekommt sogar Geld dafür (teilweise). Die Kosten dafür fallen in den Redispatch und werden am Ende wieder auf alle umgelegt, über die Netzentgelte.

In so einem Szenario wäre es doch trotzdem sinnvoll, die PV im Süden weiter laufen zu lassen.

bzw, vielleicht sollte man hier 2 Aspekte unterscheiden:
- Sinnvoll im Sinne von wirklich netzdienlich/netzunterstützend. Das bekommen wir, aufgrund fehlender Infos, nicht hin.
- sinnvoll im Sinne von für "mich" kurzfristig wirtschaftlich "sinnvoll", das bekommt man über den EPEX Preis hin. (ob das langfristig wirklich sinnvoll ist, oder aufgrund von erforderlichen Redispatchmaßnahmen langfristig ne 0-Nummer oder sogar kostspielig wird, wird sich zeigen)

kurz um:
Der Strommarkt ist doch erschreckend komplex und vielschichtig. 

Meine Idee damals war, den DPL offen laufen zu lassen (Zielwert = Maximum), solange der Strommarkt "normal" ist. Wenn der EPEX-Preis ins negative fällt, würd ich den DPL auf Zielwert 0 stellen, sodass ich nur noch das produziere, was ich aktuell benötige. Wenn der Preis inkl. Netzentgelte negativ werden würde, müsste man die PV komplett deaktivieren und Strom beziehen (natürlich nur, wenn man n stündlichen flexiblen Tarif hat)

aktuell bin ich davon wieder abgekommen. die PV läuft und speist ein, was sie kann. einen Akku habe ich noch nicht und auch keinen flexiblen Stromtarif.
Da ich sehe, dass unsere 300kW Anlage im Betrieb mit angebundenem Direktvermarkter bisher noch nicht ein mal gedrosselt wurde, stellt sich mir auch die Frage, warum ich es denn zu Hause bei so ner kleinen Anlage drosseln sollte. der Strom scheint (zumindest in meiner Region) ja benötigt zu werden. Auch wenn der EPEX Preis negativ wird. da speist sie halt am Wochenende fröhlich ihre 250kW ein, auch wenn wir Samstags und Sonntags negative Börsenpreise haben.

**Reply by 1-am-r00t** [reply]
> Du willst also die maximale Leistung des DPL dynamisch, in Abhängigkeit vom EPEX Preis, einstellen. 100% Leistung, wenn der Strompreis "maximal" ist, 0% (bzw, 10%) wenn er "minimal" ist.
> Weil: du willst keinen Strom "verschwenden", wenn der EPEX Preis negativ ist.

correct

> Wo genau bleibt da deine eigene Versorgung? Willst du den DPL drosseln, weil der Strom auf der Börse tagsüber günstig ist, und statt deine Grundlast selbst zudecken lieber einkaufen?

Yes, I'd rather "buy" some energy while it's cheap, so I can use the generated solar power from my battery when prices are high.

> Wäre es nicht sinniger, den Zielwert des DPL anzupassen, abhängig vom Börsenpreis? Sodass die eigene Grundlast doch gedeckt wird.

I think the end result might be the same for my goal. Leaving the limit "open" (i.e. maximum) will feed the grid with energy for which I'm not getting a single cent of money while the power distributors will suck every single cent from me or I'm ending up "just" exporting zero - which I'm already doing.

> Ich hatte solche Gedanken in der Vergangenheit ebenfalls. aber nicht, weil ich keinen Strom "verschwenden" oder "verschenken" will, sondern weil ich niemanden "zur Last" fallen möchte.
> Aber, der Strommarkt ist viel zu komplex und für uns, Stand heute, undurchsichtig, als das wir das alles überblicken und abbilden könnten. Dafür fehlen uns einfach viel zu viele Infos.
> [...]

Exactly, and since it's so complex and the energy suppliers / EPEX Spot are not this transparent / differentiate your local power distribution, I'm gonna continue using (or some might say "abusing") this system as long as it exists in it's current form - eventually it might change one day, who knows :) But it's honestly not my job as a customer to take into consideration that energy distribution between south and north (wind power vs. gas power redispatching) might be bad because I'm no longer feeding the grid. But this seems to be a rather philosophical discussion which we cannot influence in my opinion.

> aktuell bin ich davon wieder abgekommen. die PV läuft und speist ein, was sie kann. einen Akku habe ich noch nicht und auch keinen flexiblen Stromtarif.

The scenario/discussion I initiated is meant for people like me that have both a battery attached to their inverters AND an spot tariff (i.e. EPEX Spot Hourly). Of course, if I'd be in your position (no battery nor spot tariff), my point of view would be different probably.

**Comment by genmad** [comment]
Hi, great idea to charge your battery while energy is cheap and use it whiöe it is expensive. I believe your Description falls short on some points ( energy producing forecast, current soc, expected consumption and maybe more).

Instead of setting the maximum power by mqtt, you could also create a virtual powermeter which you connect to odob and by controlling the virtualpowermeter you can achieve your goal too.

When the price is high, then route the currently required grid power to odob, when the price is low then route only the minimum Power required to keep the inverter running to odob. And in the times inbetween you do some algorithm to come up with a good solution.

For reference here an example where up to 5 inverters are using a virtual powermeter to supply enough power independent of the power required on each phase.  https://github.com/genmad/shellyAPIGen2AndAboveScripts/blob/master/src/openDTUonBattery/virtualPowerMeters_oDoB.js

**Reply by 1-am-r00t** [reply]
Hi!

> great idea to charge your battery while energy is cheap and use it whiöe it is expensive

I am not per-se actively charging the battery using a power supply, but rather charge it "naturally" using solar power leveraging my Victron MPPT charger - just in case you misunderstood :)

>  believe your Description falls short on some points ( energy producing forecast, current soc, expected consumption and maybe more)

true that, however, I am not sure if I would like to go down that deep into this rabbit hole involving also the points you've mentioned - even if I fully agree :+1: 

> Instead of setting the maximum power by mqtt, you could also create a virtual powermeter which you connect to odob and by controlling the virtualpowermeter you can achieve your goal too.

thats an interesing approach, I haven't really considered this to be honest but I got your point. I will definitely look into it. The concept sounds good because like this I could circumvent with the currently given options adding/asking for additional functions and/or MQTT endpoints and I would not have to do any huge changes to my existing setup since I'm relying on an http-out node in Node-Red already for my power metering which I am using within odob :)

Thanks for stopping by and taking the time to share your feedback / idea :)
